### ISSUE TITLE
Compare LocalDateTime with XMLGregorianCalendar

### ISSUE THREAD

**Issue by gervaisb** [joel-costigliola/assertj-core#1127]
#### Summary
When testing object mapping, it can be useful to compare `LocalDateTime` with `XMLGregorianCalendar` without having to compare field by field or via a filed extractor

#### Example

```java
LocalDate date = // ... 2017-01-01 02:03:04
XmlGregorinaCalendar calendar = // ... 2017-01-01 02:03:04
Assertions.assertThat(date).isEqualTo(calendar); // true
```

#### Java 8 specific ?

* YES : create Pull Request from the `master` branch
~* NO  : create Pull Request from the `2.x` branch~

**Comment by joel-costigliola** [comment]
Sure, @gervaisb are you keen to contribute this one ?

**Comment by gervaisb** [comment]
Of course. As soon as I have some freetime.

I was thinking to add an new `isEqualTo` method into
`AbstractLocalDateAssert` and `AbstractLocalDateTimeAssert` to begin. Then
add the same in `AbstractOffsetTimeAssert` and
`AbstractOffsetDateTimeAssert`.

This morning I gave a look to your tests to understand how and where to add
my tests. But I have some questions.
Should I add an `XmlGregorianCalendar` field and all related methods into
`AbstractDateAssertWithDateArg_Test` or create another
`AbstractDateAssertWithXmlGregorianCalendarArg_Test` ? If I have to create
another abstract test class, what name should I use for the concrete test
class ?
To be honest, if you had time to send me a sketch of the tests classes, I
will gain a lot of time.

**Comment by joel-costigliola** [comment]
You can ask them here :)

**Comment by gervaisb** [comment]
About the tests. Should I update `AbstractDateAssertWithDateArg_Test` to
add an `XmlGregorianCalendar` or do you prefer another
`AbstractDateAssertWithXmlGregorianCalendarArg_Test` ?
Le mar. 5 déc. 2017 à 10:10, Joel Costigliola <notifications@github.com> a
écrit :

**Comment by joel-costigliola** [comment]
The first option is fine, that is adding `protected abstract DateAssert assertionInvocationWithXmlGregorinaCalendarArg();` to `AbstractDateAssertWithDateArg_Test`.

We might also rename `AbstractDateAssertWithDateArg_Test` to `AbstractDateAssertWithDifferentArgType_Test` or something alike.

**Comment by gervaisb** [comment]
I have added `isEqualTo(XMLGregorianCalendar)` to `AbstractLocalDateAssert`, `AbstractOffsetTimeAssert`, `AbstractLocalTimeAssert`, `AbstractLocalDateTimeAssert` and `AbstractOffsetDateTimeAssert`. Can you give your first feedback ?

https://github.com/gervaisb/assertj-core/tree/feature/compare-LocalDateTime-with-XMLGregorianCalendar

**Comment by joel-costigliola** [comment]
@gervaisb please create a PR, that's make it easier to review contribution.

**Comment by joel-costigliola** [comment]
I think the only non confusing comparison is `ZonedDateTime` against `XMLGregorianCalendar` since they both represents a date with a time zone.
@gervaisb what is the concrete use case, the example shown comparing a `LocalDate` with a `XMLGregorianCalendar` can work locally and fail on a CI server in a different timezone.

**Comment by gervaisb** [comment]
Well the XmlGregorianCalendar is used by jaxb (by default) to represent
either an xsd: time, xsd:datetime or xsd:date and they may have a time
zone. That's why I allowed comparison on all Java.time types. Allowing the
user comparing what he want by excepting he know what he does.
Another option is to provide an `assertEqualsWithoutFields` where we expect
a varargs of fields to ignore (day, date, time, timezones,...)

Le sam. 16 déc. 2017 à 08:28, Joel Costigliola <notifications@github.com> a
écrit :

**Comment by joel-costigliola** [comment]
Fair enough, it is then important to clearly explain what the assertion behavior is regarding to time zones, especially when comparing local date/time with a XmlGregorianCalendar.

I will add my other comments in the PR.

**Comment by filiphr** [comment]
`XmlGregorationCalendar` uses `DatatypeConstants#FIELD_UNDEFINED` for fields that are not set. I don't know what would be best to do, but one possibility would be to check that the fields which are not relevant to the java time to be `DatatypeConstants#FIELD_UNDEFINED`.

For example `LocalDate` would check the day, month and year and all the rest should be undefined.

Just my 2 cents, it doesn't have to be like that, just food for thought :)

**Comment by gervaisb** [comment]
I had those fields in mind when proposing the `assertEqualsWithoutFields`. But
regarding the future issue with Java9 and the unclear behavior when
comparing with `java.time` classes I propose to move everything into a new
`XmlGregorianCalendarAssertions
` class.
Whe must be able to move it anywhere for Java9 and the ˋisEqualToˋ method
can be overloaded for each type of `java.time` to fail if the calendar
fields does not represent the type and compare it according to the type.
If I am a bit shellfish this is better for our uses cases where the
"actual" is the XmlGregorianCalendar and the "expected" is the date.

**Comment by nach-o-man** [comment]
Greetings,

maybe it is an ok idea to introduce some generic solution like assertion method "using adapter/transform". Similar to "usingComparator".

for example: 
```java
    LocalDateTime date = LocalDateTime.of(2017, 1, 1, 2, 3, 4); // ... 2017-01-01 02:03:04
    XMLGregorianCalendar calendar = XMLGregorianCalendarImpl.createDateTime(2017, 1, 1, 2, 3, 4); // ... 2017-01-01 02:03:04
    Function<LocalDateTime, XMLGregorianCalendar> adapter = (LocalDateTime dateTime) -> XMLGregorianCalendarImpl.createDateTime(dateTime.getYear(),
                                                                                                                                dateTime.getMonthValue(),
                                                                                                                                dateTime.getDayOfMonth(),
                                                                                                                                dateTime.getHour(),
                                                                                                                                dateTime.getMinute(),
                                                                                                                                dateTime.getSecond());
    Assertions.assertThat(date).usingAdapter(adapter).isEqualTo(calendar); // true
```

I could try to implement it.

**Comment by gervaisb** [comment]
Hello. It is already what I am doing. Except that the adapter is not
managed by AssertJ.
ˋassertThat(dateTime).isEqualTo(toDateTime(xmlCalendar))`
Le mer. 28 févr. 2018 à 20:03, Andrey Kuzmin <notifications@github.com> a
écrit :

**Comment by phgachoud** [comment]
Managed with `assertThat(xmlCal.toGregorianCalendar().toZonedDateTime())
				.isCloseTo(expectedGregorianCalendar.toZonedDateTime(), byLessThan(3, ChronoUnit.SECONDS));`
